### PR TITLE
Rework Object's [[OwnPropertyKeys]]

### PR DESCRIPTION
**Pull request by szilagyiadam**
I've removed the ecma_op_object_get_property_names method, and implemented the following ones:
- ecma_op_object_own_property_keys: this is now the internal [[OwnPropertyKeys]] method
- ecma_op_object_enumerate: this is used for the for-in iterator
- ecma_object_sort_property_names: this is used for sorting the property names of an object
- ecma_object_list_lazy_property_names: this is for getting the lazy instantiated properties
- ecma_object_prop_name_is_duplicated: this is for checking if a given property is duplicated in an object

Also the for-in operation with Proxy object works with this patch, #3992 should be closed

JerryScript-DCO-1.0-Signed-off-by: Adam Szilagyi aszilagy@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 1]
```diff
@@ -1329,6 +1303,83 @@ ecma_builtin_object_dispatch_routine (uint16_t builtin_routine_id, /**< built-in
...
+    obj_p = ecma_get_object_from_value (arg1);
+    ecma_collection_t *props_p = ecma_op_object_own_property_keys (obj_p);
+
+    if (props_p->item_count == 0)
+    {
+      ecma_collection_destroy (props_p);
+      return ecma_op_create_array_object (NULL, 0, false);
+    }
+
+    ecma_value_t result_array = ecma_op_create_array_object (props_p->buffer_p, props_p->item_count, false);
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c:1377]: This is a fairly big chunk of code. We should move it into a method.

File: jerry-core/ecma/operations/ecma-objects.c [hunk 2]
```diff
@@ -2014,79 +2014,80 @@ ecma_op_object_get_enumerable_property_names (ecma_object_t *obj_p, /**< routine
...
-        ecma_collection_push_back (properties_p, ecma_copy_value (names_buffer_p[i]));
+        status = ecma_proxy_object_get_own_property_descriptor (obj_p, key_p, &prop_desc);
       }
       else
+  #endif /* ENABLED (JERRY_BUILTIN_PROXY) */
       {
-        /* 4.a.ii.2.a */
-        ecma_value_t value = ecma_op_object_get (obj_p, key_p);
+        status = ecma_op_object_get_own_property_descriptor (obj_p, key_p, &prop_desc);
+      }
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-objects.c:2034]: Do we really need the `if-else` here? The `ecma_op_object_get_own_property_descriptor` already checks internally if the input object is a Proxy or not. For now maybe we could use the `ecma_op_object_get_own_property_descriptor` in every case.

File: jerry-core/ecma/operations/ecma-objects.h [hunk 3]
```diff
@@ -67,9 +67,12 @@ ecma_value_t ecma_op_object_get_own_property_descriptor (ecma_object_t *object_p
                                                          ecma_property_descriptor_t *prop_desc_p);
 ecma_value_t ecma_op_object_has_instance (ecma_object_t *obj_p, ecma_value_t value);
 ecma_value_t ecma_op_object_is_prototype_of (ecma_object_t *base_p, ecma_object_t *target_p);
-ecma_collection_t * ecma_op_object_get_property_names (ecma_object_t *obj_p, uint32_t opts);
+// master
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-objects.h:70]: These comments seems to be irrelevant.

File: jerry-core/ecma/operations/ecma-objects.h [hunk 4]
```diff
@@ -90,7 +93,6 @@ jmem_cpointer_t ecma_op_ordinary_object_get_prototype_of (ecma_object_t *obj_p);
 ecma_value_t ecma_op_ordinary_object_set_prototype_of (ecma_object_t *base_p, ecma_value_t proto);
 bool JERRY_ATTR_PURE ecma_op_ordinary_object_is_extensible (ecma_object_t *object_p);
 void ecma_op_ordinary_object_prevent_extensions (ecma_object_t *object_p);
-
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-objects.h:93]: Please keep the empty newline here.

File: jerry-core/api/jerry.c [hunk 5]
```diff
@@ -2886,8 +2886,14 @@ jerry_get_object_keys (const jerry_value_t obj_val) /**< object value */
     return jerry_throw (ecma_raise_type_error (ECMA_ERR_MSG (wrong_args_msg_p)));
   }
 
-  return ecma_builtin_helper_object_get_properties (ecma_get_object_from_value (obj_val),
-                                                    ECMA_LIST_ENUMERABLE);
+  ecma_collection_t *prop_names = ecma_op_object_get_enumerable_property_names (ecma_get_object_from_value (obj_val),
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/api/jerry.c:2889]: It is possible.

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 6]
```diff
@@ -881,40 +836,67 @@ ecma_builtin_object_object_define_properties (ecma_object_t *obj_p, /**< routine
...
-    property_descriptors[property_descriptor_number].flags |= ECMA_PROP_IS_THROW;
+        if (ECMA_IS_VALUE_ERROR (desc_obj))
+        {
+          ecma_free_property_descriptor (&prop_desc);
+          goto cleanup;
+        }
 
-    ecma_free_value (desc_obj);
+        ecma_value_t conv_result = ecma_op_to_property_descriptor (desc_obj,
+                                                                    &property_descriptors[property_descriptor_number]);
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c:871]: misaligned line.

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 7]
```diff
@@ -881,40 +836,67 @@ ecma_builtin_object_object_define_properties (ecma_object_t *obj_p, /**< routine
...
       goto cleanup;
     }
+#endif /* ENABLED (JERRY_ESNEXT) */
 
-    /* 5.b */
-    ecma_value_t conv_result = ecma_op_to_property_descriptor (desc_obj,
-                                                               &property_descriptors[property_descriptor_number]);
+    if (ecma_is_value_true (get_desc))
+    {
+      if ((prop_desc.flags & ECMA_PROP_IS_ENUMERABLE))
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c:860]: no double `(())` is needed.

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c (hunk 1 above)
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-object.c:1377]: +1

File: jerry-core/vm/vm.c [hunk 8]
```diff
@@ -1765,7 +1765,8 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
           }
 
           ecma_object_t *object_p = ecma_get_object_from_value (result);
-          ecma_collection_t *names_p = ecma_op_object_get_property_names (object_p, ECMA_LIST_ENUMERABLE);
+          ecma_collection_t *names_p = ecma_op_object_get_enumerable_property_names (object_p,
+                                                                                     ECMA_ENUMERABLE_PROPERTY_KEYS);
```
Comment (rerobika) [inline jerry-core/vm/vm.c:1769]: misaligned line.

File: jerry-core/ecma/operations/ecma-objects.c [hunk 9]
```diff
@@ -2096,621 +2097,426 @@ ecma_op_object_get_enumerable_property_names (ecma_object_t *obj_p, /**< routine
...
+      {
+        bool is_enumerable = (prop_desc.flags & ECMA_PROP_IS_ENUMERABLE) != 0;
+        ecma_free_property_descriptor (&prop_desc);
 
-        if (insertion_pos == array_index_name_pos)
-        {
-          names_p[array_index_name_pos++] = name_p;
-        }
-        else
+        if (ecma_has_string_value_in_collection (visited_names_p, name_p)
```
Comment (galpeter) [inline jerry-core/ecma/operations/ecma-objects.c:2467]: At the moment the `ecma_has_string_value_in_collection` is guarded by the `JERRY_BUILTIN_JSON` macro. Using this here would mean that the method's header and implementation would need to move a bit. This is the reason for one of the Travis CI failure.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 10]
```diff
@@ -1220,8 +1218,7 @@ ecma_builtin_array_prototype_object_sort (ecma_value_t this_arg, /**< this argum
   JERRY_ASSERT (ecma_is_value_empty (ret_value));
 
   /* Undefined properties should be in the back of the array. */
-
-  buffer_p = array_index_props_p->buffer_p;
+  ecma_value_t *buffer_p = array_index_props_p->buffer_p;
```
[the remark this answers is not in the document]
Comment (szilagyiadam) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c:1222]: IMHO it is ok, beacuse the ```JMEM_DEFINE_LOCAL_ARRAY``` above will create a new execution block until ```JMEM_FINALIZE_LOCAL_ARRAY``` and the two ```buffer_p```  variable is not in the same block

File: tests/jerry/object-define-properties.js [hunk 11]
```diff
@@ -173,32 +173,3 @@ Object.defineProperties(obj, {
 assert (obj.a === "b");
 assert (obj.bar === 42);
 assert (obj.a === undefined);
-
-// This code should throw TypeError
```
[the remark this answers is not in the document]
Comment (szilagyiadam) [reply tests/jerry/object-define-properties.js:177]: I moved the test back and changed accordingly to the PR
